src/main/java/com/vmware/vipclient/i18n/messages/service/LocaleService.java

```diff
@@ -31,119 +34,177 @@ public LocaleService(LocaleDTO dto) {
-        if (regionMap != null) {
-            return regionMap;
+        LocaleCacheItem cacheItem = getRegionsByLocale(locale);
+        if (!cacheItem.getCachedData().isEmpty()) {
+            return cacheItem.getCachedData();
         }
         Iterator<Locale> fallbackLocalesIter = LocaleUtility.getFallbackLocales().iterator();
         while (fallbackLocalesIter.hasNext()) {
             String fallbackLocale = fallbackLocalesIter.next().toLanguageTag();
             if(fallbackLocale.equalsIgnoreCase(locale))
                 continue;
             logger.info("Can't find regions for locale [{}], look for fallback locale [{}] regions as fallback!", locale, fallbackLocale);
-            regionMap = getRegionsByLocale(fallbackLocale);
-            if (regionMap != null) {
-                new FormattingCacheService().addRegions(locale, regionMap);
+            cacheItem = getRegionsByLocale(fallbackLocale);
+            if (!cacheItem.getCachedData().isEmpty()) {
+                new FormattingCacheService().addRegions(locale, cacheItem);
                 logger.debug("Fallback locale [{}] regions is cached for locale [{}]!\n\n", fallbackLocale, locale);
                 break;
             }
         }
-        return regionMap;
+        return cacheItem.getCachedData();
     }
 
-    public Map<String, String> getRegionsByLocale(String locale){
+    public LocaleCacheItem getRegionsByLocale(String locale){
         if(locale != null && !locale.isEmpty())
             locale = locale.replace("_", "-").toLowerCase();
-        Map<String, String> regionMap = null;
+        LocaleCacheItem cacheItem = null;
         logger.debug("Look for region list from cache for locale [{}]", locale);
         FormattingCacheService formattingCacheService = new FormattingCacheService();
-        regionMap = formattingCacheService.getRegions(locale);
-        if (regionMap != null) {
+        cacheItem = formattingCacheService.getRegions(locale);
+        if (cacheItem != null) {
+            if (cacheItem.isExpired()) { // cacheItem has expired
+                // Update the cache in a separate thread
+                populateRegionsCache(locale, cacheItem);
+            }
             logger.debug("Find regions from cache for locale [{}]!", locale);
-            return regionMap;
+            return cacheItem;
         }
-        regionMap = getRegionsFromDS(locale, VIPCfg.getInstance().getMsgOriginsQueue().listIterator());
-        if (regionMap != null) {
+        cacheItem = new LocaleCacheItem();
+        getRegionsFromDS(locale, cacheItem, VIPCfg.getInstance().getMsgOriginsQueue().listIterator());
+        if (!cacheItem.getCachedData().isEmpty()) {
             logger.debug("Find the regions for locale [{}].\n", locale);
-            formattingCacheService.addRegions(locale, regionMap);
+            formattingCacheService.addRegions(locale, cacheItem);
             logger.debug("Regions is cached for locale [{}]!\n\n", locale);
```

Comment:
For locales that are not found in cache, we need to immediately return fallback locale's cacheItem first and do line 72-77 in a separate thread. Reason:

For some product applications, this slowness at the very first uncached request is not acceptable. So we give the option to initialize cache at application start up by setting initializeCache config to true. This will populate the cache for all supported locales. However, cache will not be populated for non-supported locales. Hence, for the very first uncached request for any locale (supported or not), we want to return the fallback locale's cacheItem, assuming that at the very least, the fallback locale's cacheItem is in cache.